core/optaplanner-constraint-streams-bavet/src/main/java/org/optaplanner/constraint/streams/bavet/common/AbstractGroupNode.java

```diff
@@ -46,44 +63,60 @@ public void insert(InTuple_ tuple) {
-            outTuple.setState(BavetTupleState.CREATING);
-            group = new Group<>(groupKey, resultContainer, outTuple);
-            groupMap.put(groupKey, group);
-            // Don't add it if (state == CREATING), but (group != null), which is a 2th insert of the same groupKey.
-            dirtyGroupQueue.add(group);
-        }
-        group.parentCount++;
-        Runnable undoAccumulator = hasCollector ? accumulate(group.resultContainer, tuple) : null;
-        GroupPart<Group<OutTuple_, GroupKey_, ResultContainer_>> groupPart = new GroupPart<>(group, undoAccumulator);
-        tupleStore[groupStoreIndex] = groupPart;
+        GroupKey_ groupKey = hasMultipleGroups ? groupKeyFunction.apply(tuple) : null;
+        createTuple(tuple, tupleStore, groupKey);
+    }
+
+    private void createTuple(InTuple_ tuple, Object[] tupleStore, GroupKey_ newGroupKey) {
+        Group<OutTuple_, GroupKey_, ResultContainer_> newGroup = getOrCreateGroup(newGroupKey);
+        newGroup.parentCount++;
+        Runnable undoAccumulator = hasCollector ? accumulate(newGroup.resultContainer, tuple) : null;
+        GroupPart<Group<OutTuple_, GroupKey_, ResultContainer_>> newGroupPart = new GroupPart<>(newGroup, undoAccumulator);
+        tupleStore[groupStoreIndex] = newGroupPart;
 
-        switch (group.outTuple.getState()) {
+        OutTuple_ outTuple = newGroup.outTuple;
+        switch (outTuple.getState()) {
             case CREATING:
-                break;
             case UPDATING:
                 break;
             case OK:
-                group.outTuple.setState(BavetTupleState.UPDATING);
-                dirtyGroupQueue.add(group);
+                outTuple.setState(BavetTupleState.UPDATING);
+                dirtyGroupQueue.add(newGroup);
                 break;
             case DYING:
-                group.outTuple.setState(BavetTupleState.UPDATING);
+                outTuple.setState(BavetTupleState.UPDATING);
                 break;
             case ABORTING:
-                group.outTuple.setState(BavetTupleState.CREATING);
+                outTuple.setState(BavetTupleState.CREATING);
                 break;
             case DEAD:
             default:
-                throw new IllegalStateException("Impossible state: The group (" + group + ") in node (" +
-                        this + ") is in an unexpected state (" + group.outTuple.getState() + ").");
+                throw new IllegalStateException("Impossible state: The group (" + newGroup + ") in node (" +
+                        this + ") is in an unexpected state (" + outTuple.getState() + ").");
         }
     }
 
+    private Group<OutTuple_, GroupKey_, ResultContainer_> getOrCreateGroup(GroupKey_ key) {
+        if (hasMultipleGroups) {
+            return groupMap.computeIfAbsent(key, this::createGroup);
+        } else {
+            if (singletonGroup == null) {
+                singletonGroup = createGroup(key);
```

Comment:
Motivation for lazy initialization (and reason for not doing in the constructor) would be nice to have (see other comment)